pkg/store/file_log.go

```diff
@@ -374,7 +381,75 @@ func (fl *fileLog) Stats() (LogStats, error) {
 }
 
 func (fl *fileLog) Close() error {
-	return fl.releaser.Release()
+	return nil
+}
+
+func topicPath(root, topic string) string {
+	return filepath.Join(root, topic)
+}
+
+type fileTopicLogs struct {
+	fsys              fs.Filesystem
+	root              string
+	segmentTargetSize int64
+	segmentBufferSize int64
+	reporter          EventReporter
+	logs              map[string]Log
+}
+
+// NewFileTopicLogs returns TopicLogs backed by a filesystem.
+func NewFileTopicLogs(
+	fsys fs.Filesystem,
+	root string,
+	segmentTargetSize int64,
+	segmentBufferSize int64,
+	reporter EventReporter,
+) TopicLogs {
+	return &fileTopicLogs{
+		fsys:              fsys,
+		root:              root,
+		segmentTargetSize: segmentTargetSize,
+		segmentBufferSize: segmentBufferSize,
+		reporter:          reporter,
+	}
+}
+
+func (ft *fileTopicLogs) Create(t string) (WriteSegment, error) {
+	p := topicPath(ft.root, t)
+	if !ft.fsys.Exists(p) {
+		if err := ft.fsys.MkdirAll(p); err != nil {
+			return nil, err
+		}
+	}
+	return newFileWriteSegment(ft.fsys, p)
+}
+
+func (ft *fileTopicLogs) Get(t string) (Log, error) {
+	p := topicPath(ft.root, t)
+	if !ft.fsys.Exists(p) {
+		return nil, ErrTopicNotFound
+	}
+	return NewFileLog(ft.fsys, p, ft.segmentTargetSize, ft.segmentBufferSize, ft.reporter)
+}
+
+func (ft *fileTopicLogs) All() (map[string]Log, error) {
+	fis, err := ft.fsys.ReadDir(ft.root)
+	if err != nil {
+		return nil, err
+	}
+	m := map[string]Log{}
+	for _, fi := range fis {
+		if !fi.IsDir() {
+			continue
+		}
+		t := fi.Name()
+		l, err := NewFileLog(ft.fsys, topicPath(ft.root, t), ft.segmentTargetSize, ft.segmentBufferSize, ft.reporter)
+		if err != nil {
+			return nil, err
+		}
+		m[t] = l
+	}
+	return m, nil
```

Comment:
Always instantiating new logs seems wasteful (allocs and syscalls) – but compared to when we read/write logs it probably won't matter.